master/buildbot/www/hooks/bitbucketserver.py

```diff
@@ -0,0 +1,158 @@
+                'repository': repo_url,
+                'author': '{} <{}>'.format(payload['actor']['displayName'],
+                                           payload['actor']['username']),
+                'comments': 'Bitbucket Server commit {}'.format(
+                    payload_change['new']['target']['hash']),
+                'branch': GIT_HEAD_REF.format(payload_change['new']['name']),
+                'project' : project,
+                'category' : 'push'
+            }
+
+            if callable(self._codebase):
+                change['codebase'] = self._codebase(payload)
+            elif self._codebase is not None:
+                change['codebase'] = self._codebase
+
+            changes.append(change)
+
+        return (changes, payload['repository']['scmId'])
+
+    def handle_pullrequest_created(self, payload):
+        return self.handle_pullrequest(
+            payload,
+            GIT_MERGE_REF.format(int(payload['pullrequest']['id'])),
+            "pull-created")
+
+    def handle_pullrequest_updated(self, payload):
+        return self.handle_pullrequest(
+            payload,
+            GIT_MERGE_REF.format(int(payload['pullrequest']['id'])),
+            "pull-updated")
+
+    def handle_pullrequest_fulfilled(self, payload):
+        return self.handle_pullrequest(
+            payload,
+            GIT_HEAD_REF.format(
+                payload['pullrequest']['toRef']['branch']['name']),
+            "pull-fulfilled")
+
+    def handle_pullrequest_rejected(self, payload):
+        return self.handle_pullrequest(
+            payload,
+            GIT_HEAD_REF.format(
+                payload['pullrequest']['fromRef']['branch']['name']),
+            "pull-rejected")
+
+    def handle_pullrequest(self, payload, refname, category):
+        pr_number = int(payload['pullrequest']['id'])
+        repo_url = payload['repository']['links']['self'][0]['href']
+        repo_url = repo_url.rstrip('browse')
+        change = {
+            'revision': None,
+            'revlink': payload['pullrequest']['link'],
+            'repository': repo_url,
+            'author': '{} <{}>'.format(payload['actor']['displayName'],
+                                       payload['actor']['username']),
+            'comments': 'Bitbucket Server Pull Request #{}'.format(pr_number),
+            'branch' : refname,
+            'project': payload['repository']['project']['name'],
+            'category': category,
+            'properties' : {'pullrequesturl' : payload['pullrequest']['link']}
+        }
+
+        if callable(self._codebase):
+            change['codebase'] = self._codebase(payload)
+        elif self._codebase is not None:
+            change['codebase'] = self._codebase
+
+        return [change], payload['repository']['scmId']
+
+
+def getChanges(request, options=None):
+    """
+    Process the Bitbucket Server webhook event.
+
+    :param twisted.web.server.Request request: the http request object
+
+    """
```

Comment:
You have a docstring here, is this intentional, or should this be only in the accompanying rst files?